File: ios-mip-app/FFCI/Analytics/MipAnalytics.swift
//
//  MipAnalytics.swift
//  FFCI
//
//  Centralized Firebase Analytics navigation and content events (ticket 029).
//

import FirebaseAnalytics
import Foundation

enum MipAnalytics {
    /// Matches CMS home sentinel used in menu filtering.
    static let homePageUuid = "__home__"

    /// Log when the app becomes active (not automatically collected on iOS).
    static func logAppOpen() {
        Analytics.logEvent(AnalyticsEventAppOpen, parameters: nil)
    }

    /// GA4 `screen_view` with stable `screen_name` and optional SwiftUI/view class name.
    static func logScreenView(screenName: String, screenClass: String? = nil) {
        let name = truncate(screenName, maxLen: 100)
        var params: [String: Any] = [
            AnalyticsParameterScreenName: name
        ]
        if let screenClass {
            params[AnalyticsParameterScreenClass] = truncate(screenClass, maxLen: 100)
        }
        Analytics.logEvent(AnalyticsEventScreenView, parameters: params)
    }

    /// Spec `content_view` with page identity for engagement reporting.
    static func logContentView(pageUuid: String, title: String, contentType: String) {
        Analytics.logEvent("content_view", parameters: [
            "page_uuid": truncate(pageUuid, maxLen: 100),
            "page_title": truncate(title, maxLen: 100),
            "content_type": truncate(contentType, maxLen: 100)
        ])
    }

    /// Spec `external_link` when opening Safari / system browser (ticket 031).
    static func logExternalLink(
        url: URL,
        pageUuid: String?,
        pageTitle: String?,
        linkLabel: String?,
        linkSource: String
    ) {
        var params: [String: Any] = [
            "link_url": truncate(url.absoluteString, maxLen: 100),
            "link_source": truncate(linkSource, maxLen: 100)
        ]
        if let host = url.host, !host.isEmpty {
            params["link_domain"] = truncate(host, maxLen: 100)
        }
        if let pageUuid, !pageUuid.isEmpty {
            params["page_uuid"] = truncate(pageUuid, maxLen: 100)
        }
        if let pageTitle, !pageTitle.isEmpty {
            params["page_title"] = truncate(pageTitle, maxLen: 100)
        }
        if let linkLabel, !linkLabel.isEmpty {
            params["link_label"] = truncate(linkLabel, maxLen: 100)
        }
        Analytics.logEvent("external_link", parameters: params)
    }

    private static func truncate(_ string: String, maxLen: Int) -> String {
        guard string.count > maxLen else { return string }
        return String(string.prefix(maxLen))
    }
}

File: android-mip-app/app/src/main/java/com/fiveq/ffci/analytics/MipAnalytics.kt
package com.fiveq.ffci.analytics

import android.content.Context
import android.net.Uri
import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics

object MipAnalytics {
    const val HOME_PAGE_UUID = "__home__"

    fun logAppOpen(context: Context) {
        analytics(context)?.logEvent(FirebaseAnalytics.Event.APP_OPEN, null)
    }

    fun setAnalyticsCollectionEnabled(context: Context, enabled: Boolean) {
        analytics(context)?.setAnalyticsCollectionEnabled(enabled)
    }

    fun logScreenView(
        context: Context,
        screenName: String,
        screenClass: String? = null
    ) {
        val params = Bundle().apply {
            putString(FirebaseAnalytics.Param.SCREEN_NAME, truncate(screenName))
            screenClass?.let {
                putString(FirebaseAnalytics.Param.SCREEN_CLASS, truncate(it))
            }
        }
        analytics(context)?.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, params)
    }

    fun logContentView(
        context: Context,
        pageUuid: String,
        title: String,
        contentType: String
    ) {
        val params = Bundle().apply {
            putString("page_uuid", truncate(pageUuid))
            putString("page_title", truncate(title))
            putString("content_type", truncate(contentType))
        }
        analytics(context)?.logEvent("content_view", params)
    }

    fun logExternalLink(
        context: Context,
        url: String,
        pageUuid: String? = null,
        pageTitle: String? = null,
        linkLabel: String? = null,
        linkSource: String
    ) {
        val params = Bundle().apply {
            putString("link_url", truncate(url))
            putString("link_source", truncate(linkSource))

            val host = runCatching { Uri.parse(url).host.orEmpty() }.getOrDefault("")
            if (host.isNotBlank()) {
                putString("link_domain", truncate(host))
            }
            if (!pageUuid.isNullOrBlank()) {
                putString("page_uuid", truncate(pageUuid))
            }
            if (!pageTitle.isNullOrBlank()) {
                putString("page_title", truncate(pageTitle))
            }
            if (!linkLabel.isNullOrBlank()) {
                putString("link_label", truncate(linkLabel))
            }
        }
        analytics(context)?.logEvent("external_link", params)
    }

    private fun analytics(context: Context): FirebaseAnalytics? {
        if (!isFirebaseConfigured(context)) return null
        return FirebaseAnalytics.getInstance(context.applicationContext)
    }

    private fun isFirebaseConfigured(context: Context): Boolean {
        return context.resources.getIdentifier(
            "google_app_id",
            "string",
            context.packageName
        ) != 0
    }

    private fun truncate(value: String, maxLen: Int = 100): String {
        return if (value.length <= maxLen) value else value.take(maxLen)
    }
}
